doc/src/guides/solving/solve-ode.md

```diff
@@ -0,0 +1,341 @@
+# Solve an Ordinary Differential Equation (ODE) Algebraically
+
+Use SymPy to solve an ordinary differential equation (ODE) algebraically. For
+example, solving $y''(x) + 9y(x)=0 $ yields $ y(x)=C_{1} \sin(3x)+ C_{2}
+\cos(3x)$.
+
+## Alternatives to Consider
+- To numerically solve a system of ODEs, use a [SciPy ODE
+  solver](https://docs.scipy.org/doc/scipy/reference/integrate.html#solving-initial-value-problems-for-ode-systems)
+  such as `solve_ivp`.
```

Comment:
A super common thing to do is to create an ode in sympy, lambdify it for use in solve_ivp, and then integrate numerically. I think it could be helpful for us to show that here (or another guide). We have some examples in the code generation tutorial: https://www.sympy.org/scipy-2017-codegen-tutorial/ and here is an example from my course materials: https://moorepants.github.io/learn-multibody-dynamics/simulation.html.

Reply:
Thanks, @moorepants. I would be happy to include such an example on this page. I'm having a tough time converting an [example from the code generation tutorial](https://www.sympy.org/scipy-2017-codegen-tutorial/notebooks/25-chemical-kinetics-intro.html#Solving-the-initial-value-problem-numerically) from [odeint](https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.odeint.html?highlight=odeint#scipy.integrate.odeint) (SciPy's outdated function) to [solve_ivp](https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.solve_ivp.html#scipy.integrate.solve_ivp) and using [lambdify](https://docs.sympy.org/latest/modules/utilities/lambdify.html?highlight=lambdify#module-sympy.utilities.lambdify). Here's what I've tried and the error it gives. Any assistance would be greatly appreciated.

```py
import sympy as sym
sym.init_printing()

def rhs(y, t, kf, kb):
    rf = kf * y[0]**2 * y[1]
    rb = kb * y[2]**2
    return [2*(rb - rf), rb - rf, 2*(rf - rb)]

y, k = sym.symbols('y:3'), sym.symbols('kf kb')
ydot = rhs(y, None, *k)

t = sym.symbols('t')  # not used in this case.
f = sym.lambdify((y, t) + k, ydot)  # EXERCISE: (y, t) + k

import numpy as np
from scipy.integrate import solve_ivp
tout = np.linspace(0, 10)
k_vals = np.array([0.42, 0.17])  # arbitrary in this case
y0 = [1, 1, 0]

solve_ivp(f, tout, y0, args=k_vals)
```

```py
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
/var/folders/9w/cn8tqbv976gckxd0c9cbr1sw0000gn/T/ipykernel_96785/228934301.py in <module>
----> 1 solve_ivp(f, tout, y0, args=k_vals)

~/opt/anaconda3/lib/python3.9/site-packages/scipy/integrate/_ivp/ivp.py in solve_ivp(fun, t_span, y0, method, t_eval, dense_output, events, vectorized, args, **options)
    540         method = METHODS[method]
    541 
--> 542     solver = method(fun, t0, y0, tf, vectorized=vectorized, **options)
    543 
    544     if t_eval is None:

~/opt/anaconda3/lib/python3.9/site-packages/scipy/integrate/_ivp/rk.py in __init__(self, fun, t0, y0, t_bound, max_step, rtol, atol, vectorized, first_step, **extraneous)
     92         self.max_step = validate_max_step(max_step)
     93         self.rtol, self.atol = validate_tol(rtol, atol, self.n)
---> 94         self.f = self.fun(self.t, self.y)
     95         if first_step is None:
     96             self.h_abs = select_initial_step(

~/opt/anaconda3/lib/python3.9/site-packages/scipy/integrate/_ivp/base.py in fun(t, y)
    136         def fun(t, y):
    137             self.nfev += 1
--> 138             return self.fun_single(t, y)
    139 
    140         self.fun = fun

~/opt/anaconda3/lib/python3.9/site-packages/scipy/integrate/_ivp/base.py in fun_wrapped(t, y)
     18 
     19     def fun_wrapped(t, y):
---> 20         return np.asarray(fun(t, y), dtype=dtype)
     21 
     22     return fun_wrapped, y0

~/opt/anaconda3/lib/python3.9/site-packages/scipy/integrate/_ivp/ivp.py in <lambda>(t, x, fun)
    512         # additional parameters.  Pass in the original fun as a keyword
    513         # argument to keep it in the scope of the lambda.
--> 514         fun = lambda t, x, fun=fun: fun(t, x, *args)
    515         jac = options.get('jac')
    516         if callable(jac):

<lambdifygenerated-4> in _lambdifygenerated(_Dummy_26, t, kf, kb)
      1 def _lambdifygenerated(_Dummy_26, t, kf, kb):
----> 2     [y0, y1, y2] = _Dummy_26
      3     return [2*kb*y2**2 - 2*kf*y0**2*y1, kb*y2**2 - kf*y0**2*y1, -2*kb*y2**2 + 2*kf*y0**2*y1]

TypeError: cannot unpack non-iterable float object
```

Reply:
This runs (but double check I didn't change the math).

solve_ivp takes t, y not y, t as args and the time is given as a tuple of (t0, tf) not an array. They changed several things about the api compared to odeint, which is a bit confusing.

```python
import sympy as sym
sym.init_printing()

def rhs(t, y, kf, kb):
    rf = kf * y[0]**2 * y[1]
    rb = kb * y[2]**2
    return [2*(rb - rf), rb - rf, 2*(rf - rb)]

y, k = sym.symbols('y:3'), sym.symbols('kf kb')
ydot = rhs(None, y, *k)

t = sym.symbols('t')  # not used in this case.
f = sym.lambdify((t, y) + k, ydot)  # EXERCISE: (y, t) + k

import numpy as np
from scipy.integrate import solve_ivp
k_vals = np.array([0.42, 0.17])  # arbitrary in this case
y0 = [1, 1, 0]

solve_ivp(f, (0, 10), y0, args=k_vals)
```

Reply:
This removes the k values unpacking, maybe a bit clearer for learners.

```python
import sympy as sym
sym.init_printing()

def rhs(t, y, kf, kb):
    rf = kf * y[0]**2 * y[1]
    rb = kb * y[2]**2
    return [2*(rb - rf), rb - rf, 2*(rf - rb)]

y, (kf, kb) = sym.symbols('y:3'), sym.symbols('kf kb')
ydot = rhs(None, y, kf, kb)  # show that ydot is symbolic here

t = sym.symbols('t')  # not used in this case.
f = sym.lambdify((t, y, kf, kb), ydot)

import numpy as np
from scipy.integrate import solve_ivp
k_vals = np.array([0.42, 0.17])  # arbitrary in this case
y0 = [1, 1, 0]

solve_ivp(f, (0, 10), y0, args=k_vals)
```

Reply:
Great, thanks!